WIP: Port to Win32 and MSVC

mauve opened the pull request:
Hi,

I really like your library, however I would like to use it for a project which currently requires windows support. So here is a quick WIP-version of a windows port.

I am planning to do some more cleanup before this is ready to merge but sofar it looks good, however before I spend too much time cleaning up my branch I first wanted to check if you are willing to accept a windows port.

The blocker for merging (IMHO) is that focus handling seems broken in the sense that I cannot move between checkboxes (or list items) in any of the examples. I ran the `print_key_press` example and pressed ⬅⬆➡⬇, this was the output I got for that:

```
( 27 91 68 ) -> (special)
( 27 91 65 ) -> (special)
( 27 91 67 ) -> (special)
( 27 91 66 ) -> (special)
```

- [ ] Bug: Focus handling not working
   - in the examples (e.g. `checkbox.cpp`) I can toggle the individual checkboxes but I cannot move between items, I tried to understand the focus implementation but am I unsure which keypresses would move focus between different components
- [x] Refactor: event translation into separate file
   - currently the *Console API* event mapping is living inside the `read_char` thread lambda in `screen_interactive.cpp`, as this code is a bit longer than the original (which is just a oneliner) my plan is to move it to a separate function in a separate file or ifdef guarded inside `event.cpp`
- [ ] Annoyance: Flickering
   - `screen_interactive.cpp` redraws the screen which causes the output to flicker (on Windows at least), I was thinking about only redrawing the screen if it has changed (i.e. dirty tracking)

Screenshot:

![image](https://user-images.githubusercontent.com/795800/77254989-a93b8100-6c64-11ea-9ed9-c248fe77d708.png)

mauve commented on src/ftxui/component/screen_interactive.cpp:
```diff
@@ -147,8 +147,12 @@ void ScreenInteractive::Loop(Component* component) {
+#ifdef WIN32
+  // TODO: port
+#else
   auto old_sigwinch_handler = std::signal(SIGWINCH, OnResize);
```
I am unsure how this is handled on Windows but I think just handling the `WINDOW_BUFFER_SIZE_EVENT` is the way to implement this.

mauve replied:
this is now fixed

mauve commented:
I rebased and removed the WIP commits so that the history now looks nice.

ArthurSonzogni commented:
Thanks! I am looking at this now...

ArthurSonzogni commented on .gitignore:
```diff
@@ -0,0 +1 @@
+build
```
I never use a .gitignore.
I try to be thoughful when adding files
Could you revert this change?

mauve replied:
I will revert it.

I mostly add it so that ‘git status’ only shows relevant stuff.

ArthurSonzogni commented on CMakeLists.txt:
```diff
@@ -4,6 +4,12 @@ project(ftxui
   VERSION 0.1
 )
 
+set(CMAKE_CXX_STANDARD 17)
```
I don't want to force users to user C++17.
What is the reason you want to add this?

mauve replied:
Will fix.

Doesn’t this do the same thing as target_compile_features(... PUBLIC cxx_std_17) below btw?

ArthurSonzogni replied:
I haven't seen I am already using C++17. Sorry please ignore my comment.

ArthurSonzogni commented on CMakeLists.txt:
```diff
@@ -4,6 +4,12 @@ project(ftxui
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+
+add_compile_options("$<$<C_COMPILER_ID:MSVC>:/utf-8>")
```
What about using target_compile_option(<target> PRIVATE "your option")
This avoid affecting other targets the user of the library might have.

mauve replied:
Will fix, this was a remainder from my first try.

ArthurSonzogni commented on CMakeLists.txt:
```diff
@@ -81,7 +81,9 @@ foreach(lib screen dom component)
   )
   target_compile_features(${lib} PUBLIC cxx_std_17)
-  target_compile_options(${lib} PRIVATE -Wall -Werror -pedantic -Wextra -Wno-sign-compare)
+  if(NOT MSVC)
```
I am not against this change, but could you explain why it is needed?

mauve replied:
MSVC doesn’t understand those command line arguments, it uses warnings levels /W[1-4] (or you can enable or disable individual warnings using /W[ed]XXXX)

ArthurSonzogni replied:
Okay it looks good to me.

ArthurSonzogni commented on src/ftxui/dom/border.cpp:
```diff
@@ -1,6 +1,8 @@
 #include "ftxui/dom/node.hpp"
 #include "ftxui/dom/elements.hpp"
 
+#include <algorithm>
```
I don't see why this is needed.
Could you please remove every non necessary includes?

mauve replied:
It is required though, it might be that std::min/max is accidentally pulled in through some other std include, however that is not the same across implementations so for msvc it is needed.

ArthurSonzogni replied:
I see! Thanks!

ArthurSonzogni commented on .vscode/c_cpp_properties.json:
```diff
@@ -0,0 +1,22 @@
+{
+    "configurations": [
```
I have limited understanding of how vscode is working.
Could you explain why we need this?

mauve replied:
I can remove all the .vscode/ files if you want to.

This particular file is kinda auto generated by the c++ plugin for vscode. Sadly it contains some real settings and some generated ones.

Anyway if you want I can remove all .vscode/ files.

ArthurSonzogni replied:
Perfect! Yes, I would like not having this file.

mauve commented:
What clang-format settings do you use? Would you mind if I put them in a .clang-format file, that would allow me to enable format on save.

ArthurSonzogni commented:
> What clang-format settings do you use? Would you mind if I put them in a .clang-format file, that would allow me to enable format on save.

There is already a .clang-format:
~~~
# Defines the Chromium style for automatic reformatting.
# http://clang.llvm.org/docs/ClangFormatStyleOptions.html
BasedOnStyle: Chromium
Standard: Cpp11
~~~

ArthurSonzogni commented:
I have issues compiling on Linux after you patch:
~~~cpp
/home/sonzogna/programmation/real/FTXUI/src/ftxui/component/event_input_listener.cpp:158:1: error: expected unqualified-id before ‘while’
  158 | while (!quit_) {
      | ^~~~~
/home/sonzogna/programmation/real/FTXUI/src/ftxui/component/event_input_listener.cpp:164:1: error: expected declaration before ‘}’ token
  164 | }  // namespace ftxui
      | ^
make[2]: *** [CMakeFiles/component.dir/build.make:115: CMakeFiles/component.dir/src/ftxui/component/event_input_listener.cpp.o] Error 1
make[2]: *** Attente des tâches non terminées....
~~~

ArthurSonzogni commented on include/ftxui/component/event_input_listener.hpp:
```diff
@@ -0,0 +1,55 @@
+//
+// On Windows systems:
+//   - Uses ReadConsoleInput for resize and keypresses
+class EventInputListener {
```
Why do we need this class?

Can't we spawn some thread listening for event and use PostEvent for synchronization with the main thread?

mauve replied:
I started out by having just a single thread and a lambda in `ScreenInteractive` but it just became too crowded, also especially since we actually need two threads.

The two threads are required in order to be able to reuse `Event::GetEvent` for input parsing. I was quickly considering turning the parser used by `Event::GetEvent` into a push-parser rather than pull-parser but my pull-request started becoming too big.

Thread 1: is used for waiting and dequeuing `INPUT_RECORD`s using `ReadConsoleInput`, all high-level events are immediate posted (via `consumer`), all characters are put on an internal character stream queue.

Thread 2: (this thread is also used on *NIX)
  - on NIX: uses `::getchar()`
  - on Win: pulls (blockingly) chars from the character stream queue mentioned above

mauve replied:
> I have issues compiling on Linux after you patch:
> 
> ```c++
> /home/sonzogna/programmation/real/FTXUI/src/ftxui/component/event_input_listener.cpp:158:1: error: expected unqualified-id before ‘while’
>   158 | while (!quit_) {
>       | ^~~~~
> /home/sonzogna/programmation/real/FTXUI/src/ftxui/component/event_input_listener.cpp:164:1: error: expected declaration before ‘}’ token
>   164 | }  // namespace ftxui
>       | ^
> make[2]: *** [CMakeFiles/component.dir/build.make:115: CMakeFiles/component.dir/src/ftxui/component/event_input_listener.cpp.o] Error 1
> make[2]: *** Attente des tâches non terminées....
> ```

fixed in latest commit

ArthurSonzogni commented:
> The blocker for merging (IMHO) is that focus handling seems broken in the sense that I cannot move between checkboxes (or list items) in any of the examples. I ran the `print_key_press` example and pressed ⬅⬆➡⬇, this was the output I got for that:
> 
> ```
> ( 27 91 68 ) -> (special)
> ( 27 91 65 ) -> (special)
> ( 27 91 67 ) -> (special)
> ( 27 91 66 ) -> (special)
> ```
> 
> * [ ]  Bug: Focus handling not working
>   
>   * in the examples (e.g. `checkbox.cpp`) I can toggle the individual checkboxes but I cannot move between items, I tried to understand the focus implementation but am I unsure which keypresses would move focus between different components

I am getting the same things on linux.
They must correspond to what we have in components/event.cpp:
~~~cpp
Event Event::ArrowLeft = Event::Special("\x1B[D");
Event Event::ArrowRight = Event::Special("\x1B[C");
Event Event::ArrowUp = Event::Special("\x1B[A");
Event Event::ArrowDown = Event::Special("\x1B[B");
~~~

I am surprised it doesn't for on Windows for you.

ArthurSonzogni commented:
By the way, thank you very much for adding support for Windows. I was hoping it could it yet. That's awesome!

I will try to add a travis CI to ensuire it compile no matter the OS after every patches.

ArthurSonzogni commented:
I did some updates in the repository.
- Added Travis. This way, you can check on every operating system if your change is compiling. I expect Travis to be triggered for pull requests. https://travis-ci.org/github/ArthurSonzogni/FTXUI
- Run clang format on every files. I added many clang-format off for the sections that do require not to be updated.
- I tried to make it compile a Window as far a possible using the travis output. There are still many errors.

mauve commented:
Hmm got a Mac compile error which I quickly fixed, but really what year is it?

![image](https://user-images.githubusercontent.com/795800/77288101-61097680-6cd7-11ea-88dd-02b6fbbc9ae2.png)

mauve commented:
> > What clang-format settings do you use? Would you mind if I put them in a .clang-format file, that would allow me to enable format on save.
> 
> There is already a .clang-format:
> 
> ```
> # Defines the Chromium style for automatic reformatting.
> # http://clang.llvm.org/docs/ClangFormatStyleOptions.html
> BasedOnStyle: Chromium
> Standard: Cpp11
> ```

Ok wonder how I missed it. Now that we have nested preprocessor statements you might want consider to enable preprocessor indentation, for example by adding `IndentPPDirectives: AfterHash`, which would turn this:

(garbled mess)
```cpp
#ifdef _WIN32
#define WIN32_LEAN_AND_MEAN
#define NOMINMAX
#include <Windows.h>

#include <deque>

#ifndef UNICODE
#error Must be compiled in UNICODE mode
#endif
#else
#include <csignal>
#endif
```

(readable mess)
```cpp
#ifdef _WIN32
#  define WIN32_LEAN_AND_MEAN
#  define NOMINMAX
#  include <Windows.h>

#  include <deque>


#  ifndef UNICODE
#    error Must be compiled in UNICODE mode
#  endif
#else
#  include <csignal>
#endif
```

mauve commented:
Noijs, build is now fixed. Unsure about that Google Build stuff, it wants me to login which I am not really willing to do.

ArthurSonzogni commented:
> Ok wonder how I missed it. Now that we have nested preprocessor statements you might want consider to enable preprocessor indentation, for example by adding `IndentPPDirectives: AfterHash`, which would turn this:

Yes, I think I would like it.

ArthurSonzogni commented:
I would like to make a few changes to your CLs. I started landing your patches with some modifications one by one.
- https://github.com/ArthurSonzogni/FTXUI/commit/ec34842b54d524609c79cb33d085fbbed43532a5
- ...

I am not very sure about the last one about he InputEventListener. I need to take a deeper look. Would you mind waiting a bit until I merged everything except this one?

What email address do you want to use? The one you used for those commit where not associated to your github account. I chose to use one associated to you github account instead. Please let me know if you don't want to use this address.

mauve commented:
> What email address do you want to use? The one you used for those commit where not associated to your github account. I chose to use one associated to you github account instead. Please let me know if you don't want to use this address.

Thanks, I accidentily used my work-email.

I rebased on `upstream/master` and added the `IndentPPDirectives` statement.

mauve commented:
> I am not very sure about the last one about he InputEventListener. I need to take a deeper look. Would you mind waiting a bit until I merged everything except this one?

I think you can make general input handling working using just normal `::getchar` on windows (not super sure though, did not test it), however for resize handling and mouse events etc you need to use `ReadConsoleInput()` which requires the extra-dance, FYI.

mauve commented:
I don't really know what I am supposed to do now, I keep rebasing on master and you keep force-pushing it.

Do you want me to keep rebasing my PR?

Some or your latest changes introduced some interesting merge conflicts. Also this piece of code you added (https://github.com/ArthurSonzogni/FTXUI/blob/master/src/ftxui/component/screen_interactive.cpp#L115-L117) is not necessary, the console mode is a per-process setting AFAIK (it just sets the mode on the console i/o buffer, not on any object shared between processes or the terminal).

```cpp
  on_exit_functions.push([original_mode]() {
    SetConsoleMode(GetStdHandle(STD_OUTPUT_HANDLE), original_mode);
  });
```

mauve commented:
> I am getting the same things on linux.
> They must correspond to what we have in components/event.cpp:
> 
> ```c++
> Event Event::ArrowLeft = Event::Special("\x1B[D");
> Event Event::ArrowRight = Event::Special("\x1B[C");
> Event Event::ArrowUp = Event::Special("\x1B[A");
> Event Event::ArrowDown = Event::Special("\x1B[B");
> ```
> 
> I am surprised it doesn't for on Windows for you.

Can you point me to the thing which is responsible for focus-handling? I seem to be too stupid to find it myself.

ArthurSonzogni commented:
> > I am getting the same things on linux.
> > They must correspond to what we have in components/event.cpp:
> > ```c++
> > Event Event::ArrowLeft = Event::Special("\x1B[D");
> > Event Event::ArrowRight = Event::Special("\x1B[C");
> > Event Event::ArrowUp = Event::Special("\x1B[A");
> > Event Event::ArrowDown = Event::Special("\x1B[B");
> > ```
> > 
> > 
> > I am surprised it doesn't for on Windows for you.
> 
> Can you point me to the thing which is responsible for focus-handling? I seem to be too stupid to find it myself.

Yes. Every components can define children and elect an "ActiveChild".
An element is Focused when all of its ancestors are "ActiveChild"
See https://github.com/ArthurSonzogni/FTXUI/blob/b6d1f8ebdbb49fcb71359a5be7ed30beda0dd92b/include/ftxui/component/component.hpp#L31

Components can define arbitrary behaviors. One subclass "Container" defines a reasonable way to handle events in:
https://github.com/ArthurSonzogni/FTXUI/blob/b6d1f8ebdbb49fcb71359a5be7ed30beda0dd92b/src/ftxui/component/container.cpp#L32
~~~cpp
bool Container::OnEvent(Event event) {
  if (!Focused())
    return false;

  if (ActiveChild()->OnEvent(event))
    return true;

  return (this->*event_handler_)(event);
}
~~~
Whenever the user try to navigate: First, forward the event to the active children. If the children is able to handle the event it returns true. Second, (If it can't), then its parent try to handle the event and select a different ActiveChild.

ArthurSonzogni commented:
> I don't really know what I am supposed to do now, I keep rebasing on master and you keep force-pushing it.
> 
> Do you want me to keep rebasing my PR?

Sorry about this. There are many different patches. I was thinking I could keep&edit your patches one by ones without bothering you with many comments.

To avoid you all the rebases, can I force-push on your branch everytime one commit is accepted? (there is nothing to do or you)

ArthurSonzogni commented:
I pushed on your branch. All your commits have been merged, except the last two ones. I am still looking at them.

ArthurSonzogni commented:
I am now understanding way more what event_input_listener was doing. That's actually very good!

I will make a few changes and try to land this tomorrow.

For the second patch, do you know if there is a MSVC option to disable the unsigned char comparisons warnings? If possible, I would like to avoid the second change with all the static_cast.

mauve commented:
In general in MSVC you can disable any warning in different ways:

1. `#pragma warning`
1. `/Wd[warning]` (suppress warning)
1. `/W[level][warning]` (remap warning to `[level]`)
1. `/We[warning]` (turn warning to error)

In this case you want `/Wd4018` I guess - there might be others as well I don't remember at the top of my head (refer: https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-3-c4018?view=vs-2019).

The warning number is part of the output/error message.

mauve commented:
![image](https://user-images.githubusercontent.com/795800/77413812-43640c00-6dc0-11ea-8373-3137ad467660.png)

This is from stepping around in the debugger in `checkbox.exe`: `selector_` is pointing to some garbage memory.

ArthurSonzogni commented:
I completely merged all you patches for now. Thank you very much!

I added the Sender/Receiver primitives. The Event parser is now working like a coroutine. It allowed me to simplify a bit your patch and transform the class EventInputLIstener into a simple function.
I disabled warning 4244 for now.

I don't have a Windows operating system. I can't say if my patch was working the way it was intended or not. It compiles at least:
https://travis-ci.org/github/ArthurSonzogni/FTXUI/jobs/666599799
Please let me know if this is working.

I will try tomorrow to look at your problem with "focus". Thanks for the dbg informations you provided.

ArthurSonzogni commented:
I made some new test and added support for running them on Windows:
https://travis-ci.com/github/ArthurSonzogni/FTXUI/jobs/320259295

We can see what you talked about:
~~~
Bug: Focus handling not working
in the examples (e.g. checkbox.cpp) I can toggle the individual checkboxes but I cannot move between items, I tried to understand the focus implementation but am I unsure which keypresses would move focus between different components
~~~

The test doesn't pass:
ContainerTest.HorizontalEvent

ArthurSonzogni commented:
I understand the bug.

I am doing in the class declaration.
~~~
  int selected_ = 0;
  int* selector_ = &selected_;
~~~

Then when we move the instance, the |selector_| point to an invalid address.

This is working on non-Windows platform, because the compiler are smart enough. They are using what is called: "return value optimization".

I will submit a fix today.
